### PR TITLE
Uphold/StopWhenUnneeded/BindsTo: add retry timer on rate limit

### PR DESCRIPTION
**Pull request by bluca**
The Upholds= promise is that as long as unit A is up and Upholds=B, B will be activated if failed or inactive. But there is a hard-coded, non-configurable rate limit for this, so if B has a dependency on C, and C fails, B will never be restarted.

Requeue a unit when its job finishes, and also requeue after the interval has passed when hitting the rate limit.
Apply to BindsTo= and StopWhenUnneeded= as well.

### CODE REVIEW

File: src/core/unit.c [hunk 1]
```diff
@@ -517,52 +517,58 @@ void unit_add_to_dbus_queue(Unit *u) {
         u->in_dbus_queue = true;
 }
 
-void unit_submit_to_stop_when_unneeded_queue(Unit *u) {
-        assert(u);
+int unit_submit_to_stop_when_unneeded_queue(sd_event_source *s, uint64_t usec, void *userdata) {
```
Comment (poettering) [inline src/core/unit.c]: nah, please don't.

please have single handler, that will add the unit to all three queues. When the rate limit window is hit this might have effect on more than one of the queues. And adding to the queues is "free", hence better add to to many than to too few...

Comment (poettering) [reply src/core/unit.c]: in particular as we only maintain a single ratelimiter for all three queues. so if we hit the limit for one of the queues then we'll also hit it for the others, but then we won't check them again when the window is over.

File: src/core/manager.c [hunk 2]
```diff
@@ -1389,6 +1389,29 @@ static unsigned manager_dispatch_gc_job_queue(Manager *m) {
...
+
+        assert(u);
+        assert(cb);
+
+        if (ratelimit_below(&u->auto_start_stop_ratelimit))
+                return 1;
+
+        r = sd_event_add_time_relative(
+                        u->manager->event,
+                        &u->auto_start_stop_event_source,
```
Comment (poettering) [inline src/core/manager.c:1422]: if auto_start_stop_event_source is already initialized you'll leak an event source here.

File: src/core/manager.c [hunk 3]
```diff
@@ -1389,6 +1389,29 @@ static unsigned manager_dispatch_gc_job_queue(Manager *m) {
...
+                        CLOCK_MONOTONIC,
+                        u->auto_start_stop_ratelimit.interval,
```
Comment (poettering) [inline src/core/manager.c]: please add:

```
static inline usec_t ratelimit_end(const RateLimit *rl) {
        assert(rl); 
        if (rl->begin == 0)
                return 0;
        return usec_add(rl->begin, rl->interval);
}
```

to ratelimit.h

and then use an absolute event source and use that for the absolute time to retry again.

this has the benefit that we'll try the instant the window is over, and not any later.

File: src/core/job.c [hunk 4]
```diff
@@ -1051,6 +1051,8 @@ int job_finish_and_invalidate(Job *j, JobResult result, bool recursive, bool alr
                         job_add_to_gc_queue(other->job);
                 }
 
+        (void) unit_submit_to_start_when_upheld_queue(NULL, 0, u);
```
Comment (poettering) [inline src/core/job.c]: this should be a separate commit, no? also, please add a brief comment why this is useful here

Comment (poettering) [reply src/core/job.c]: hmm, this shouldn't add `u` to the queue, but all units that come after it. it should be moved inside the UNIT_ATOM_BEFORE loop immediately preceeding here, and operate on `other`

File: src/core/manager.c (hunk 2 above)
Comment (poettering) [reply src/core/manager.c:1422]: you should either reuse the existing event source and restart it if it already elapsed (i.e. is disabled – but leave it on if it is still pending, and exit early here).

Or we should unref/disable the event source whenever it elapses from its callback handler.

**Comment by poettering** [comment]
i am not maintaining the stable tree, but I'd not consider this a bugfix, but an extension o behaviour.

**Comment by bluca** [comment]
> i am not maintaining the stable tree, but I'd not consider this a bugfix, but an extension o behaviour.

The documentation for Upholds= says:

```
Configures dependencies similar to Wants=, but as long as this unit is up, all units listed in Upholds=
are started whenever found to be inactive or failed, and no job is queued for them. While a Wants=
dependency on another unit has a one-time effect when this units started, a Upholds= dependency on it
has a continuous effect, constantly restarting the unit if necessary.
```

To me, this is not met without this change, hence why I'd consider it a bug fix?

**Comment by bluca** [comment]
> > To me, this is not met without this change, hence why I'd consider it a bug fix?
> 
> dunno, it's a corner case I'd say: if things persistently are fucked. previous strategy: "if the thing just keeps dying we give up". with your PR the strategy is "if the thing just keeps dyning we pause for a bit". I think both strategies in this corner case are valid. i can see you prefer the latter and I agree, but the comment so far makes no hint about the corner case at all, hence both strategies are totally within the boundaries of the docs.
> 
> docs never go into all detail of the code. for example ratelimiting of /proc/self/mountinfo events we don't document either.

ok, label dropped, ptal

File: src/core/job.c [hunk 5]
```diff
@@ -1051,6 +1051,10 @@ int job_finish_and_invalidate(Job *j, JobResult result, bool recursive, bool alr
                         job_add_to_gc_queue(other->job);
                 }
 
+        /* Ensure that when an upheld unit activation job fails we requeue it, if it still needs to be
+         * uphold. If there are no state changes in the uphelder, it would not be retried otherwise. */
+        unit_submit_to_start_when_upheld_queue(u);
```
Comment (poettering) [inline src/core/job.c:1056]: as in the earlier comment: this should be inside the UNIT_FOREACH_DEPENDENCY() loop above, and operate on `other`, not on `u`, no?

**Comment by poettering** [comment]
first commit looks great. second commit looks fishy

File: src/core/job.c (hunk 5 above)
Comment (bluca) [reply src/core/job.c:1056]: I don't think so? It works this way: A upholds B, B fails because of a dep, B's job here fails so B queues itself up again. I think we do want this.
We already have requeue on state changes of A, so it shouldn't be necessary to do it again here, no?

Comment (poettering) [reply src/core/job.c:1056]: waiting for a job is one thing. propagating failure from one job to another is another thing. I am pretty sure that we should recheck the Upholds= stuff also in the case where we delay one job for another, but do not propagate the failure.

I do think adding the unit itself (as in your current patch) to the queue probably also makes sense though.

And actually I figure we probably should add the units to all three queues again.

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
@bluca 
Not sure, but maybe this causes https://bugzilla.redhat.com/show_bug.cgi?id=2208864.

**Comment by bluca** [comment]
Is Fedora using Uphold= for those?

**Comment by yuwata** [comment]
No. After reconsidering the issue, it seems unrelated to this PR. Sorry for the noise.
